R/dice.R

```diff
@@ -147,76 +155,44 @@ dice <- function(x, fm = NULL, SPC = TRUE, pctMissing = FALSE, fill = FALSE, str
   )
   
   # convert to DT
-  if(! inherits(h, 'data.table')) {
+  if (!inherits(h, 'data.table')) {
     h <- as.data.table(h)
   }
   
-  ## TODO: are keys worth the extra sorting / re-sorting?
-  # init keys, sorts the data on hzID (alpha-sort)
-  # setkeyv(h, hzidn)
-  # consider an index, seems to have no effect
-  # setindexv(h, hzidn)
-  
-  ## expand 1 unit slices to max depth of each profile
-  # TODO: this will have to be made more intelligent in the presence of overlap
-  # mapply() call takes 1/4 of total time
-  
-  # safe wrapper to base::seq
-  # NA in from/to (hz depths) not allowed
-  # TODO: from == to (0-thickness) horizons ??
-  .seq.safe <- function(from, to, by) {
-    if(any(is.na(from)) | any(is.na(to))) {
-      stop('corrupt horizonation, consider using `strict = TRUE`', call. = FALSE)
-    }
-    
-    return(
-      seq(from = from, to = to, by = by)
-    )
+  # from == to (0-thickness) horizons -- these should be removed -AGB 2022/08/17
+  zthk <- h[[htb[1]]] == h[[htb[2]]]
+  if (any(zthk, na.rm = TRUE)) {
+    message("horizons with zero thickness have been omitted from results")
   }
   
-  tops <- mapply(
-    FUN = .seq.safe, 
-    from = h[[htb[1]]], 
-    to = h[[htb[2]]] - 1, 
-    by = 1, 
-    SIMPLIFY = FALSE
-  )
+  h.sub <- as.data.table(h[which(!zthk),])
   
-  tops <- unlist(tops)
+  # expand 1 unit slices to max depth of each profile
```

Comment:
Is this more efficient than the previous version? If it is, then I'd like to expand it out a little bit for readability.

Reply:
I'm all for it, maybe we can look at it together so I can better understand exactly what is happening in there. I'll just expand out a little for visual clarity. Robust > simplicity here.